scripts/bundle-images.sh

```diff
@@ -0,0 +1,245 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# bundle-images.sh bundles all of the images used across HMC into a single
+# tarball.  This is useful for deploying HMC into air-gapped environments.
+# It is recommended to use this script in conjunction with the Makefile target
+# `make bundle-images` which will perform some additional steps outside of
+# this scripts functionality.
+# Usage: make bundle-images
+set -x
+
+if [ "$YQ" == "" ] ||
+    [ "$IMG" == "" ] ||
+    [ "$KUBECTL" == "" ] ||
+    [ "$HELM" == "" ] ||
+    [ "$KIND_CLUSTER_NAME" == "" ] ||
+    [ "$NAMESPACE" == "" ] ||
+    [ "$BUNDLE_TARBALL" == "" ] ||
+    [ "$EXTENSIONS_BUNDLE_TARBALL" == "" ] ||
+    [ "$TEMPLATES_DIR" == "" ];
+    then
+        echo "This script should not be run directly.  Use 'make bundle-images' instead."
+        exit 1
+fi
+
+LABEL_KEY="cluster.x-k8s.io/provider"
+IMAGES_BUNDLED="$IMG"
+EXTENSION_IMAGES_BUNDLED=""
+
+echo -e "\nBundling images for HMC, this may take awhile...\n"
+
+trap ctrl_c INT
+
+function wait_for_deploy_exist() {
+    local deployment_label=$1
+    local max_wait_secs=300
+    local interval_secs=5
+    local start_time
+
+
+    start_time=$(date +%s)
+
+    echo "Verifying provider Deployment with label: \"$deployment_label\" exists in namespace: \"$NAMESPACE\"..."
+
+    while true; do
+        current_time=$(date +%s)
+            if (( (current_time - start_time) > max_wait_secs )); then
+                echo "Error: Waited for Deployment with label: \"$deployment_label\" in namespace: \"$NAMESPACE\" to exist for $max_wait_secs seconds and it still does not exist."
+                return 1
+            fi
+
+        output=$($KUBECTL -n "$NAMESPACE" get deploy -l $deployment_label)
+
+        if [[ output != "" ]]; then
+            echo "Deployment in namespace: \"$NAMESPACE\" with label: \"$deployment_label\" exists."
+            break
+        else
+            sleep $interval_secs
+        fi
+    done
+}
+
+function bundle_images() {
+    local images=$1
+    local tarball=$2
+
+    echo "Bundling images into $tarball..."
+    docker save -o $tarball $images
+    if [[ $? -ne 0 ]]; then
+        echo "Error: Failed to bundle images into $tarball"
+        exit 1
+    fi
+}
+
+
+function ctrl_c() {
+        echo "Caught CTRL-C, exiting..."
+        exit 1
+}
+
+echo -e "\nVerifying provider Deployments are ready...\n"
+
+# Verify each provider we support has deployed so we can get the images used
+# across the deployments.
+for TEMPLATE in $(find $TEMPLATES_DIR -name 'provider.yaml');
+do
+    RESULT=$(perl -nle 'print if m{(?=.*Provider)(?=.*kind)}' $TEMPLATE)
```

Comment:
What if you'll get?
```bash
$ perl
-bash: perl: command not found
```
It looks like it's a job for awk here

Also there is a size limit on variable content.

Reply:
`awk` also has different implementations, mawk, gawk,...

Reply:
Right, so it's tricky to choose the right ones, we could just suggest on mac to install `ggrep` from Brew prior to running the script and just use that, but it's the same as `perl`, it's hard to satisfy every requirement.

Reply:
One thing that we have on all platforms is Go, how about writing all scripts in Go?

Reply:
I think conversation went a little bit to another direction.
1. I'm not aware of the requirement to run release pipeline (which this script will be part of) on MacOS (this is how you get endless scope)
2. Let's stick to GNU utilities, the MacOS quirks (if any) could be simply documented in `dev.md`. 
3. Go is an option, but I guess it's too late to make this decisions at this point. We can do followup task, but it should be evaluated separately